{{cookiecutter.project_name}}/.github/workflows/test.yaml

```diff
@@ -18,44 +18,61 @@ defaults:
     shell: bash -euo pipefail {0}
 
 jobs:
+  get-environments:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.get-envs.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          filter: blob:none
+          fetch-depth: 0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+      - name: Get test environments
+        id: get-envs
+        run: |
+          MATRIX_JSON=$(uvx hatch env show --json | jq -c 'to_entries
+            | map(
+                select(.key | startswith("hatch-test"))
+                | {
+                    env_name: .key,
+                    name: (if (.key | contains("pre")) then .key + " (PRE-RELEASE DEPENDENCIES)" else .key end),
+                    python: .value.python
+                  }
+              )')
+          echo "matrix=${MATRIX_JSON}" | tee $GITHUB_OUTPUT
+
   test:
+    needs: get-environments
     runs-on: ${{ matrix.os }}
 
     strategy:
       fail-fast: false
       matrix:
-        include:
-          - os: ubuntu-latest
-            python: "3.10"
-          - os: ubuntu-latest
-            python: "3.12"
-          - os: ubuntu-latest
-            python: "3.12"
-            pip-flags: "--pre"
-            name: PRE-RELEASE DEPENDENCIES
-
-    name: ${{ matrix.name }} Python ${{ matrix.python }}
-
-    env:
-      OS: ${{ matrix.os }}
-      PYTHON: ${{ matrix.python }}
+        os: [ubuntu-latest]
+        include: ${{ fromJson(needs.get-environments.outputs.matrix) }}
+
+    name: ${{ matrix.name }}
 
     steps:
       - uses: actions/checkout@v4
-        with:
-          filter: blob:none
-          fetch-depth: 0
       - name: Install uv
         uses: astral-sh/setup-uv@v5
         with:
           cache-dependency-glob: pyproject.toml
+      - name: create hatch environment
+        run: |
+          uvx hatch python install ${{ matrix.python }}
+          uvx hatch env create ${{ matrix.env_name }}
       - name: run tests using hatch
         env:
           MPLBACKEND: agg
           PLATFORM: ${{ matrix.os }}
           DISPLAY: :42
-        run: uvx hatch test --cover --python ${{ matrix.python }}
+        run: uvx hatch run ${{ matrix.env_name }}:run-cov
```

Comment:
Apparently this requires to manually create the environment before, while `uvx hatch test` does not.